beacon-chain/verification/blob.go

```diff
@@ -0,0 +1,263 @@
+package verification
+
+import (
+	"context"
+
+	"github.com/pkg/errors"
+	"github.com/prysmaticlabs/prysm/v4/beacon-chain/state"
+	fieldparams "github.com/prysmaticlabs/prysm/v4/config/fieldparams"
+	"github.com/prysmaticlabs/prysm/v4/config/params"
+	"github.com/prysmaticlabs/prysm/v4/consensus-types/blocks"
+	"github.com/prysmaticlabs/prysm/v4/encoding/bytesutil"
+	"github.com/prysmaticlabs/prysm/v4/runtime/logging"
+	"github.com/prysmaticlabs/prysm/v4/time/slots"
+	log "github.com/sirupsen/logrus"
+)
+
+const (
+	RequireBlobIndexInBounds Requirement = iota
+	RequireSlotBelowMaxDisparity
+	RequireSlotBelowFinalized
+	RequireValidProposerSignature
+	RequireSidecarParentSeen
+	RequireSidecarParentValid
+	RequireSidecarParentSlotLower
+	RequireSidecarDescendsFromFinalized
+	RequireSidecarInclusionProven
+	RequireSidecarBlobCommitmentProven
+	RequireSidecarProposerExpected
+)
+
+// GossipSidecarRequirements defines the set of requirements that BlobSidecars received on gossip
+// must satisfy in order to upgrade an ROBlob to a VerifiedROBlob.
+var GossipSidecarRequirements = []Requirement{
+	RequireBlobIndexInBounds,
+	RequireSlotBelowMaxDisparity,
+	RequireSlotBelowFinalized,
+	RequireValidProposerSignature,
+	RequireSidecarParentSeen,
+	RequireSidecarParentValid,
+	RequireSidecarParentSlotLower,
+	RequireSidecarDescendsFromFinalized,
+	RequireSidecarInclusionProven,
+	RequireSidecarBlobCommitmentProven,
+	RequireSidecarProposerExpected,
+}
+
+var (
+	ErrBlobInvalid                  = errors.New("blob failed verification")
+	ErrBlobIndexInBounds            = errors.Wrap(ErrBlobInvalid, "incorrect blob sidecar index")
+	ErrSlotBelowMaxDisparity        = errors.Wrap(ErrBlobInvalid, "slot is too far in the future")
+	ErrSlotBelowFinalized           = errors.Wrap(ErrBlobInvalid, "slot <= finalized checkpoint")
+	ErrValidProposerSignature       = errors.Wrap(ErrBlobInvalid, "proposer signature could not be verified")
+	ErrSidecarParentSeen            = errors.Wrap(ErrBlobInvalid, "parent root has not been seen")
+	ErrSidecarParentValid           = errors.Wrap(ErrBlobInvalid, "parent block is not valid")
+	ErrSidecarParentSlotLower       = errors.Wrap(ErrBlobInvalid, "blob slot <= parent slot")
+	ErrSidecarDescendsFromFinalized = errors.Wrap(ErrBlobInvalid, "blob parent is not descended from the finalized block")
+	ErrSidecarInclusionProven       = errors.Wrap(ErrBlobInvalid, "sidecar inclusion proof verification failure")
+	ErrSidecarBlobCommitmentProven  = errors.Wrap(ErrBlobInvalid, "sidecar kzg commitment proof verification failed")
+	ErrSidecarProposerExpected      = errors.Wrap(ErrBlobInvalid, "sidecar was not proposed by the expected proposer_index")
```

Comment:
My thinking with these names was:
- method name matches the wording of the gossip verification rule (ie they spell out what is valid, not what is invalid)
- Requirement names matches the method
- Error matches the requirement name

The result is that the errors read sort of inverted (`ErrSidecarProposerExpected` means the proposer was not expected). Is that confusing or does it make enough sense in context?

Reply:
(note that I ultimately changed the error names to be inverse of the requirements, because it was more confusing for the name to resemble a success condition)